server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -192,6 +197,73 @@ public Weight createWeight(Query query, ScoreMode scoreMode, float boost) throws
         }
     }
 
+    @Override
+    protected LeafSlice[] slices(List<LeafReaderContext> leaves) {
+        return computeSlices(leaves, queueSizeBasedExecutor.threadPoolExecutor.getPoolSize(), MINIMUM_DOCS_PER_SLICE);
+    }
+
+    /**
+     * Each computed slice contains at least 10% of the total data in the leaves with a
+     * minimum given by the <code>minDocsPerSlice</code> parameter and the final number
+     * of {@link LeafSlice} will be equal or lower than the number of available threads.
+     */
+    /* pkg private for testing */
+    static LeafSlice[] computeSlices(List<LeafReaderContext> leaves, int numThreads, int minDocsPerSlice) {
+        // total number of documents to be searched
+        final int numDocs = leaves.stream().mapToInt(l -> l.reader().maxDoc()).sum();
```

Comment:
just a question, would it make sense to think about the number of documents as a result of a query/filter.count() and drop slices ahead of time in the case of a match none? I am just trying to think about cases where decisions can be made much more efficient like with a sampling run of a cheap query. We don't need to add this now but for the future it might be something we wanna expose?